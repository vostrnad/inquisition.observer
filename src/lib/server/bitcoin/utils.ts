export const isP2WPKH = (script: Buffer): boolean => {
  return script.length === 22 && script[0] === 0x00 && script[1] === 0x14
}

export const isP2WSH = (script: Buffer): boolean => {
  return script.length === 34 && script[0] === 0x00 && script[1] === 0x20
}
