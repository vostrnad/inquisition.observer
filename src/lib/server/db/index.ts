import { MikroORM } from '@mikro-orm/better-sqlite'
import config from './mikro-orm.config'

let database: MikroORM | undefined
let databasePromise: Promise<MikroORM> | undefined

export const getDatabase = async (): Promise<MikroORM> => {
  if (databasePromise) return databasePromise

  if (!database) {
    databasePromise = (async () => {
      const db = await MikroORM.init(config)
      await db.schema.updateSchema()
      return db
    })()
    database = await databasePromise
    databasePromise = undefined
  }

  return database
}

export const initDatabase = async (): Promise<void> => {
  await getDatabase()
}

export const closeDatabaseIfConnected = async (): Promise<void> => {
  const db = await (databasePromise ?? database)
  database = undefined
  await db?.close()
}
