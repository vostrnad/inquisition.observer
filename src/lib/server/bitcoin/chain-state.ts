import { wrap } from '@mikro-orm/core'
import { config } from '$lib/server/config'
import { getDatabase } from '$lib/server/db'
import { Block } from '$lib/server/db/entities/block'
import { bufferToHex, hexToBuffer } from '$lib/server/utils/buffer'
import {
  BitcoinChainClient,
  type ChainTip,
  type Block as RpcBlock,
} from './bitcoin-chain-client'
import { getTransactionEntity } from './processing'

export class ChainState extends BitcoinChainClient {
  constructor() {
    super({
      host: config.bitcoin.rpcHost,
      auth: config.bitcoin.rpcAuth,
      zmqpubhashblock: config.bitcoin.zmqpubhashblock,
    })
  }

  protected async processBlocks(blocks: RpcBlock[]): Promise<void> {
    const db = await getDatabase()
    const em = db.em.fork()

    for (const block of blocks) {
      const dbBlock = wrap(new Block()).assign({
        height: block.height,
        hash: hexToBuffer(block.hash),
        time: block.time,
      })

      em.persist(dbBlock)

      for (let i = 0; i < block.tx.length; i++) {
        const tx = block.tx[i]
        const transactionEntity = getTransactionEntity(tx, block.height, i)

        if (transactionEntity) {
          transactionEntity.block = dbBlock
          em.persist(transactionEntity)
        }
      }
    }

    await em.flush()
  }

  protected async getStoredBlockByHeight(
    height: number,
  ): Promise<ChainTip | undefined> {
    const db = await getDatabase()
    const block = await db.em.fork().getRepository(Block).findByHeight(height)
    if (!block) return
    return this.dbBlockToChainTip(block)
  }

  protected async getStoredBlockByHash(
    hash: string,
  ): Promise<ChainTip | undefined> {
    const db = await getDatabase()
    const block = await db.em.fork().getRepository(Block).findByHash(hash)
    if (!block) return
    return this.dbBlockToChainTip(block)
  }

  protected async getStoredChainTip(): Promise<ChainTip | undefined> {
    const db = await getDatabase()
    const bestBlock = await db.em.fork().getRepository(Block).findBestBlock()
    if (!bestBlock) return
    return this.dbBlockToChainTip(bestBlock)
  }

  protected async rollbackStoredChain(chainTip: ChainTip): Promise<void> {
    const db = await getDatabase()
    await db.em.fork().getRepository(Block).deleteAboveHeight(chainTip.height)
  }

  private dbBlockToChainTip(block: Block): ChainTip {
    return {
      hash: bufferToHex(block.hash),
      height: block.height,
    }
  }
}
