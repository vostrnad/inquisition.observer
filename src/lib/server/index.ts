/* eslint-disable no-console */
import { ChainState } from './bitcoin/chain-state'
import { getDatabase } from './db'

export const startBackgroundProcess = (): void => {
  if (import.meta.env.DEV) {
    if (process.DEV_BACKGROUND_PROCESS_STARTED) {
      console.warn(
        'Background process is already running, restart the dev server to update it',
      )
      return
    }
    process.DEV_BACKGROUND_PROCESS_STARTED = true
  }

  console.log('starting server')

  const chainState = new ChainState()

  chainState.on('syncstart', () => {
    console.log('sync start')
  })

  chainState.on('syncend', () => {
    console.log('sync end')
  })

  chainState.on('chaintipupdate', (chainTip) => {
    console.log('new chain tip:', chainTip.height)
  })

  chainState.on('error', (e) => {
    console.error('chain state error:', e)
  })

  chainState.start()

  process.on('SIGINT', () => {
    chainState.stop()
    void getDatabase().then(async (db) => db.close())
  })
}
