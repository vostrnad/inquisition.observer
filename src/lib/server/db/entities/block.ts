import {
  Entity,
  EntityRepositoryType,
  Index,
  Property,
  Unique,
} from '@mikro-orm/core'
import { BlockRepository } from '$lib/server/db/repositories/block'
import { CustomBaseEntity } from './base'

@Entity({ repository: () => BlockRepository })
export class Block extends CustomBaseEntity {
  [EntityRepositoryType]?: BlockRepository

  @Property({ type: 'blob' })
  @Unique()
  @Index()
  public hash: Buffer

  @Property({ type: 'integer' })
  @Unique()
  @Index()
  public height: number

  @Property({ type: 'integer' })
  public time: number
}
