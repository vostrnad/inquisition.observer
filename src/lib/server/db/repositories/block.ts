import { EntityRepository } from '@mikro-orm/core'
import { Block } from '$lib/server/db/entities/block'
import { hexToBuffer } from '$lib/server/utils/buffer'

export class BlockRepository extends EntityRepository<Block> {
  public async findByHash(hash: Buffer | string): Promise<Block | null> {
    if (typeof hash === 'string') {
      hash = hexToBuffer(hash)
    }
    return this.findOne({ hash })
  }

  public async findByHeight(height: number): Promise<Block | null> {
    return this.findOne({ height })
  }

  public async findBestBlock(): Promise<Block | null> {
    return (
      (await this.find({}, { orderBy: { height: 'desc' }, limit: 1 })).at(0) ||
      null
    )
  }

  public async deleteAboveHeight(height: number): Promise<number> {
    return this.nativeDelete({ height: { $gt: height } })
  }
}
