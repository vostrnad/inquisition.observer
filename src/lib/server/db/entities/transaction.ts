import {
  Collection,
  Entity,
  EntityRepositoryType,
  Index,
  ManyToOne,
  OneToMany,
  Property,
  Unique,
} from '@mikro-orm/core'
import { TransactionRepository } from '$lib/server/db/repositories/transaction'
import { CustomBaseEntity } from './base'
import { Block } from './block'
import { TransactionInput } from './transaction-input'

@Entity({ repository: () => TransactionRepository })
@Unique({ properties: ['block', 'blockPosition'] })
export class Transaction extends CustomBaseEntity {
  [EntityRepositoryType]?: TransactionRepository

  @Property({ type: 'blob' })
  @Unique()
  public txid: Buffer
  @Property({ type: 'integer' })
  public vsize: number
  @Property({ type: 'integer' })
  public blockPosition: number

  @Property({ type: 'boolean' })
  @Index()
  public hasApo: boolean
  @Property({ type: 'boolean' })
  @Index()
  public hasCtv: boolean
  @Property({ type: 'boolean' })
  @Index()
  public hasCat: boolean

  @ManyToOne(() => Block, { deleteRule: 'cascade' })
  public block: Block

  @OneToMany(() => TransactionInput, 'transaction', {
    orderBy: { inputIndex: 'asc' },
  })
  public inputs = new Collection<TransactionInput>(this)
}
