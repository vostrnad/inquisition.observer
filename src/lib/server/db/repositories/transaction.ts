import { EntityRepository, type FilterQuery } from '@mikro-orm/better-sqlite'
import { Transaction } from '$lib/server/db/entities/transaction'
import { hexToBuffer } from '$lib/server/utils/buffer'

export interface PaginationOptions {
  page: number
  pageSize: number
}

export interface FilterOptions {
  filter?: FilterQuery<Transaction>
  pagination: PaginationOptions
}

export class TransactionRepository extends EntityRepository<Transaction> {
  public async findByTxid(txid: Buffer | string): Promise<Transaction | null> {
    if (typeof txid === 'string') {
      txid = hexToBuffer(txid)
    }
    return this.findOne({ txid }, { populate: ['*'] })
  }

  public async findAllPaginated(
    options: FilterOptions,
  ): Promise<[Transaction[], number]> {
    return this.findAndCount(options.filter || {}, {
      populate: ['*'],
      orderBy: [{ block: 'desc' }, { blockPosition: 'desc' }],
      limit: options.pagination.pageSize,
      offset: options.pagination.page * options.pagination.pageSize,
    })
  }
}
