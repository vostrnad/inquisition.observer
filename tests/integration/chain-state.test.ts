import { wrap } from '@mikro-orm/core'
import { describe, expect, it } from 'vitest'
import type { ChainTip } from '$lib/server/bitcoin/bitcoin-chain-client'
import { ChainState } from '$lib/server/bitcoin/chain-state'
import type { Block as RpcBlock } from '$lib/server/bitcoin/json-rpc'
import { getDatabase } from '$lib/server/db'
import { Block } from '$lib/server/db/entities/block'
import { Transaction } from '$lib/server/db/entities/transaction'
import { hexToBuffer } from '$lib/server/utils/buffer'
import { createArray } from '$lib/utils/array'
import {
  createBareInput,
  createBlock,
  createP2SHInput,
  createP2TRInput,
  createP2WSHInput,
  createTransaction,
  createWrappedP2SHInput,
} from '$tests/utils/bitcoin'

const getAllTransactions = async () => {
  const db = await getDatabase()
  const em = db.em.fork()
  const transactionRepository = em.getRepository(Transaction)
  return transactionRepository.findAll({ populate: ['*'] })
}

const getAllBlocks = async () => {
  const db = await getDatabase()
  const em = db.em.fork()
  const blockRepository = em.getRepository(Block)
  return blockRepository.findAll()
}

const insertBlocks = async (count: number) => {
  const db = await getDatabase()
  const em = db.em.fork()
  await em.persistAndFlush(
    createArray(count, () => createBlock([])).map((block, i) =>
      wrap(new Block()).assign({
        height: i,
        hash: hexToBuffer(block.hash),
        time: block.time,
      }),
    ),
  )
}

describe('ChainState', () => {
  describe('processBlocks', () => {
    const processBlocks = async (blocks: RpcBlock[]) => {
      // eslint-disable-next-line @typescript-eslint/dot-notation
      await new ChainState()['processBlocks'](blocks)
    }

    it('should process an empty array of blocks', async () => {
      await processBlocks([])

      const blocks = await getAllBlocks()
      expect(blocks).toHaveLength(0)

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(0)
    })

    it('should process an empty block', async () => {
      await processBlocks([createBlock([])])

      const blocks = await getAllBlocks()
      expect(blocks).toHaveLength(1)
      expect(blocks).toMatchSnapshot()

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(0)
    })

    it('should process blocks with CTV', async () => {
      await processBlocks([
        createBlock([
          createTransaction([createBareInput('b3')]),
          createTransaction([createP2SHInput('b3')]),
        ]),
        createBlock([
          createTransaction([createWrappedP2SHInput('b3')]),
          createTransaction([createP2WSHInput('b3')]),
          createTransaction([createP2TRInput('b3')]),
        ]),
      ])

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(5)
      expect(transactions).toMatchSnapshot()
    })

    it('should process blocks with APO', async () => {
      await processBlocks([
        createBlock([
          createTransaction([createP2TRInput('51ac')]),
          createTransaction([
            createP2TRInput(
              '21010000000000000000000000000000000000000000000000000000000000000000ac',
            ),
          ]),
        ]),
        createBlock([
          createTransaction([createP2TRInput('51ad')]),
          createTransaction([
            createP2TRInput(
              '21010000000000000000000000000000000000000000000000000000000000000000ad',
            ),
          ]),
        ]),
        createBlock([
          createTransaction([createP2TRInput('51ba')]),
          createTransaction([
            createP2TRInput(
              '21010000000000000000000000000000000000000000000000000000000000000000ba',
            ),
          ]),
        ]),
      ])

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(6)
      expect(transactions).toMatchSnapshot()
    })

    it('should process a block with CAT', async () => {
      await processBlocks([
        createBlock([createTransaction([createP2TRInput('7e')])]),
      ])

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(1)
      expect(transactions).toMatchSnapshot()
    })

    it('should process a block with CSFS', async () => {
      await processBlocks([
        createBlock([createTransaction([createP2TRInput('cc')])]),
      ])

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(1)
      expect(transactions).toMatchSnapshot()
    })

    it('should process a block with IKEY', async () => {
      await processBlocks([
        createBlock([createTransaction([createP2TRInput('cb')])]),
      ])

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(1)
      expect(transactions).toMatchSnapshot()
    })

    it('should correctly handle the annex', async () => {
      const input = createP2TRInput('7e')
      input.txinwitness.push('50')
      await processBlocks([createBlock([createTransaction([input])])])

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(1)
      expect(transactions).toMatchSnapshot()
    })

    it('should ignore inputs that do not use soft forks', async () => {
      await processBlocks([
        createBlock([
          createTransaction([
            // APO outside of P2TR
            createBareInput('51ac'),
            createP2SHInput('51ac'),
            createWrappedP2SHInput('51ac'),
            createP2WSHInput('51ac'),
            // CAT outside of P2TR
            createBareInput('7e'),
            createP2SHInput('7e'),
            createWrappedP2SHInput('7e'),
            createP2WSHInput('7e'),
            // every opcode except CTV and CAT
            createP2TRInput(
              '004f505152535455565758595a5b5c5d5e5f606162636465666768696a6b6c6d6e6f707172737475767778797a7b7c7d7f808182838485868788898a8b8c8d8e8f909192939495969798999a9b9c9d9e9fa0a1a2a3a4a5a6a7a8a9aaabacadaeafb0b1b2b4b5b6b7b8b9baff',
            ),
            // data pushes that should not be mistaken for opcodes
            createP2TRInput('01b3'),
            createP2TRInput('0251ac'),
            createP2TRInput('017e'),
          ]),
        ]),
      ])

      const blocks = await getAllBlocks()
      expect(blocks).toHaveLength(1)

      const transactions = await getAllTransactions()
      expect(transactions).toHaveLength(0)
    })
  })

  describe('getStoredBlockByHeight', () => {
    const getStoredBlockByHeight = async (height: number) => {
      // eslint-disable-next-line @typescript-eslint/dot-notation
      return new ChainState()['getStoredBlockByHeight'](height)
    }

    it('should return stored blocks by height', async () => {
      await insertBlocks(16)
      await expect(getStoredBlockByHeight(0)).resolves.toMatchSnapshot()
      await expect(getStoredBlockByHeight(5)).resolves.toMatchSnapshot()
      await expect(getStoredBlockByHeight(15)).resolves.toMatchSnapshot()
      await expect(getStoredBlockByHeight(16)).resolves.toBeUndefined()
    })
  })

  describe('getStoredBlockByHash', () => {
    const getStoredBlockByHash = async (hash: string) => {
      // eslint-disable-next-line @typescript-eslint/dot-notation
      return new ChainState()['getStoredBlockByHash'](hash)
    }

    it('should return stored blocks by hash', async () => {
      await insertBlocks(16)
      await expect(
        getStoredBlockByHash(
          '0000000000000000000000000000000000000000000000000000000000000000',
        ),
      ).resolves.toMatchSnapshot()
      await expect(
        getStoredBlockByHash(
          '0000000000000000000000000000000000000000000000000000000000000005',
        ),
      ).resolves.toMatchSnapshot()
      await expect(
        getStoredBlockByHash(
          '000000000000000000000000000000000000000000000000000000000000000f',
        ),
      ).resolves.toMatchSnapshot()
      await expect(
        getStoredBlockByHash(
          '0000000000000000000000000000000000000000000000000000000000000010',
        ),
      ).resolves.toBeUndefined()
    })
  })

  describe('getStoredChainTip', () => {
    const getStoredChainTip = async () => {
      // eslint-disable-next-line @typescript-eslint/dot-notation
      return new ChainState()['getStoredChainTip']()
    }

    it('should return the stored chaintip', async () => {
      await insertBlocks(32)
      await expect(getStoredChainTip()).resolves.toEqual({
        hash: '000000000000000000000000000000000000000000000000000000000000001f',
        height: 31,
      })
    })

    it('should return undefined when there are no blocks', async () => {
      await expect(getStoredChainTip()).resolves.toBeUndefined()
    })
  })

  describe('rollbackStoredChain', () => {
    const rollbackStoredChain = async (chainTip: ChainTip) => {
      // eslint-disable-next-line @typescript-eslint/dot-notation
      return new ChainState()['rollbackStoredChain'](chainTip)
    }

    it('should delete blocks above the given height', async () => {
      await insertBlocks(32)
      await expect(getAllBlocks()).resolves.toHaveLength(32)

      await rollbackStoredChain({
        hash: '000000000000000000000000000000000000000000000000000000000000000f',
        height: 15,
      })
      await expect(getAllBlocks()).resolves.toHaveLength(16)
    })
  })
})
