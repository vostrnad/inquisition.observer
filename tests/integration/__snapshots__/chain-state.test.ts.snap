// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChainState > getStoredBlockByHash > should return stored blocks by hash 1`] = `
{
  "hash": "0000000000000000000000000000000000000000000000000000000000000000",
  "height": 0,
}
`;

exports[`ChainState > getStoredBlockByHash > should return stored blocks by hash 2`] = `
{
  "hash": "0000000000000000000000000000000000000000000000000000000000000005",
  "height": 5,
}
`;

exports[`ChainState > getStoredBlockByHash > should return stored blocks by hash 3`] = `
{
  "hash": "000000000000000000000000000000000000000000000000000000000000000f",
  "height": 15,
}
`;

exports[`ChainState > getStoredBlockByHeight > should return stored blocks by height 1`] = `
{
  "hash": "0000000000000000000000000000000000000000000000000000000000000000",
  "height": 0,
}
`;

exports[`ChainState > getStoredBlockByHeight > should return stored blocks by height 2`] = `
{
  "hash": "0000000000000000000000000000000000000000000000000000000000000005",
  "height": 5,
}
`;

exports[`ChainState > getStoredBlockByHeight > should return stored blocks by height 3`] = `
{
  "hash": "000000000000000000000000000000000000000000000000000000000000000f",
  "height": 15,
}
`;

exports[`ChainState > processBlocks > should correctly handle the annex 1`] = `
[
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": false,
    "hasCat": true,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": false,
        "hasCat": true,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CAT",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000000",
    "vsize": 100,
  },
]
`;

exports[`ChainState > processBlocks > should process a block with CAT 1`] = `
[
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": false,
    "hasCat": true,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": false,
        "hasCat": true,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CAT",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000000",
    "vsize": 100,
  },
]
`;

exports[`ChainState > processBlocks > should process a block with CSFS 1`] = `
[
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": true,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": true,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CHECKSIGFROMSTACK",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000000",
    "vsize": 100,
  },
]
`;

exports[`ChainState > processBlocks > should process a block with IKEY 1`] = `
[
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": true,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": true,
        "inputIndex": 0,
        "scriptAsm": "OP_INTERNALKEY",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000000",
    "vsize": 100,
  },
]
`;

exports[`ChainState > processBlocks > should process an empty block 1`] = `
[
  {
    "hash": "0000000000000000000000000000000000000000000000000000000000000000",
    "height": 12,
    "time": 7200,
  },
]
`;

exports[`ChainState > processBlocks > should process blocks with APO 1`] = `
[
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": true,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": true,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_1 OP_CHECKSIG",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000000",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 2,
    "hasApo": true,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": true,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_PUSHBYTES_33 010000000000000000000000000000000000000000000000000000000000000000 OP_CHECKSIG",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000001",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000001",
      "height": 13,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": true,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": true,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_1 OP_CHECKSIGVERIFY",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000003",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000001",
      "height": 13,
      "time": 7200,
    },
    "blockPosition": 2,
    "hasApo": true,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": true,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_PUSHBYTES_33 010000000000000000000000000000000000000000000000000000000000000000 OP_CHECKSIGVERIFY",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000004",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000002",
      "height": 14,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": true,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": true,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_1 OP_CHECKSIGADD",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000006",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000002",
      "height": 14,
      "time": 7200,
    },
    "blockPosition": 2,
    "hasApo": true,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": false,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": true,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": false,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_PUSHBYTES_33 010000000000000000000000000000000000000000000000000000000000000000 OP_CHECKSIGADD",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000007",
    "vsize": 100,
  },
]
`;

exports[`ChainState > processBlocks > should process blocks with CTV 1`] = `
[
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": true,
    "hasIkey": false,
    "inputs": [
      {
        "address": null,
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": true,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CHECKTEMPLATEVERIFY",
        "type": "bare",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000000",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "height": 12,
      "time": 7200,
    },
    "blockPosition": 2,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": true,
    "hasIkey": false,
    "inputs": [
      {
        "address": "3p2sh",
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": true,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CHECKTEMPLATEVERIFY",
        "type": "p2sh",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000001",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000001",
      "height": 13,
      "time": 7200,
    },
    "blockPosition": 1,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": true,
    "hasIkey": false,
    "inputs": [
      {
        "address": "3p2sh",
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": true,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CHECKTEMPLATEVERIFY",
        "type": "p2sh-p2wsh",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000003",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000001",
      "height": 13,
      "time": 7200,
    },
    "blockPosition": 2,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": true,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1qp2wsh",
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": true,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CHECKTEMPLATEVERIFY",
        "type": "p2wsh",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000004",
    "vsize": 100,
  },
  {
    "block": {
      "hash": "0000000000000000000000000000000000000000000000000000000000000001",
      "height": 13,
      "time": 7200,
    },
    "blockPosition": 3,
    "hasApo": false,
    "hasCat": false,
    "hasCsfs": false,
    "hasCtv": true,
    "hasIkey": false,
    "inputs": [
      {
        "address": "bc1pp2tr",
        "hasApo": false,
        "hasCat": false,
        "hasCsfs": false,
        "hasCtv": true,
        "hasIkey": false,
        "inputIndex": 0,
        "scriptAsm": "OP_CHECKTEMPLATEVERIFY",
        "type": "p2tr",
      },
    ],
    "txid": "0000000000000000000000000000000000000000000000000000000000000005",
    "vsize": 100,
  },
]
`;
